backend/src/main/java/fk/prof/backend/http/BackendHttpVerticle.java

```diff
@@ -117,13 +136,55 @@ private void handlePostProfile(RoutingContext context) {
+      Recorder.ProcessGroup processGroup = RecorderProtoUtil.mapRecorderInfoToProcessGroup(pollReq.getRecorderInfo());
+      ProcessGroupContextForPolling processGroupContextForPolling = this.processGroupDiscoveryContext.getProcessGroupContextForPolling(processGroup);
+      if(processGroupContextForPolling == null) {
+        throw new IllegalArgumentException("Process group " + RecorderProtoUtil.processGroupCompactRepr(processGroup) + " not associated with the backend");
+      }
+
+      Recorder.WorkAssignment nextWorkAssignment = processGroupContextForPolling.receivePoll(pollReq);
+      Recorder.PollRes.Builder pollResBuilder = Recorder.PollRes.newBuilder()
+          .setControllerVersion(backendVersion)
+          .setControllerId(Ints.fromByteArray(ipAddress.getBytes("UTF-8")))
+          .setLocalTime(nextWorkAssignment == null
+              ? LocalDateTime.now(Clock.systemUTC()).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
+              : nextWorkAssignment.getIssueTime());
+      Recorder.PollRes pollRes;
+      if(nextWorkAssignment == null) {
+        pollRes = pollResBuilder.build();
+      } else {
+        pollRes = pollResBuilder.setAssignment(nextWorkAssignment).build();
+        AggregationWindow aggregationWindow = aggregationWindowDiscoveryContext.getAssociatedAggregationWindow(nextWorkAssignment.getWorkId());
+        if (aggregationWindow == null) {
+          throw new IllegalArgumentException(String.format("workId=%d not found, cannot associate recorder info with aggregated profile. aborting send of work assignment",
+              nextWorkAssignment.getWorkId()));
+        }
+        aggregationWindow.updateRecorderInfo(nextWorkAssignment.getWorkId(), pollReq.getRecorderInfo());
+      }
+      context.response().end(ProtoUtil.buildBufferFromProto(pollRes));
+    } catch (Exception ex) {
+      HttpFailure httpFailure = HttpFailure.failure(ex);
+      HttpHelper.handleFailure(context, httpFailure);
+    }
+  }
+
+  // /association API is requested over ELB, routed to some backend which in turns proxies it to a leader
   private void handlePutAssociation(RoutingContext context) {
     String leaderIPAddress = verifyLeaderAvailabilityOrFail(context.response());
     if (leaderIPAddress != null) {
       try {
-        //Deserialize to proto message to catch payload related errors early
-        Recorder.ProcessGroup processGroup = Recorder.ProcessGroup.parseFrom(context.getBody().getBytes());
-        //TODO: Check if backend already knows about this process group and send self as the association, rather than proxying to leader
+        Recorder.ProcessGroup processGroup = ProtoUtil.buildProtoFromBuffer(Recorder.ProcessGroup.parser(), context.getBody());
+        ProcessGroupContextForPolling processGroupContextForPolling = this.processGroupDiscoveryContext.getProcessGroupContextForPolling(processGroup);
+        if(processGroupContextForPolling != null) {
```

Comment:
Fixed in commit https://github.com/Flipkart/fk-prof/commit/4082bc7ca78e11f75aa121c72aef00de8f64c539